File: lg1726.cpp
#include <vector>
#include <cstdio>
#include <algorithm>
#include <cstring>

using namespace std;

int n, m;
vector<int> e[5007], et[5007];

void dfs(int x, vector<int> e[], bool vis[], vector<int>& a) {
    vis[x] = true;
    for (auto v : e[x])
        if (!vis[v])
            dfs(v, e, vis, a);
    a.push_back(x);
}

int main() {
    scanf("%d %d", &n, &m);
    for (int a,b,t,i=1; i<=m; ++i) {
        scanf("%d %d %d", &a, &b, &t);
        if (t==1) 
            e[a].push_back(b),
            et[b].push_back(a);
        else {
            e[a].push_back(b), e[b].push_back(a);
            et[a].push_back(b), et[b].push_back(a);
        }
    }

    vector<int> ord;
    bool vis[n+1];
    memset(vis, false, sizeof(vis));
    for (int i=1; i<=n; ++i)
        if (!vis[i]) 
            dfs(i, e, vis, ord);

    memset(vis, false, sizeof(vis));
    vector<int> ans;
    for (int i=ord.size()-1; i>=0; --i)
        if (!vis[ord[i]]) {
            vector<int> cc;
            dfs(ord[i], et, vis, cc);
            sort(cc.begin(), cc.end());
            if (cc.size()>ans.size() || (cc.size()==ans.size() && cc<ans)) 
                ans = move(cc);
        }
    printf("%d\n", (int)ans.size());
    for (auto x : ans) printf("%d ", x);

    return 0;
}

File: lg1726.java
import java.util.*;

public class lg1726 {

    public static void dfs(int x, List<ArrayList<Integer>> e, 
                           boolean[] vis, ArrayList<Integer> a) {
        vis[x] = true;
        for (int v : e.get(x))
            if (!vis[v])
                dfs(v, e, vis, a);
        a.add(x);
    }

    public static boolean smaller(ArrayList<Integer> a, ArrayList<Integer> b) {
        for (int i=0; i<a.size(); ++i)
            if (a.get(i)<b.get(i))
                return true;
            else if (a.get(i)>b.get(i))
                return false;
            return false;
    }

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);

        int n, m;
        n = in.nextInt();
        m = in.nextInt();

        List<ArrayList<Integer>> e  = new ArrayList<ArrayList<Integer>>(n+1);
        List<ArrayList<Integer>> et = new ArrayList<ArrayList<Integer>>(n+1);

        for (int i=0; i<=n; ++i) {
            e.add(new ArrayList<Integer>());
            et.add(new ArrayList<Integer>());
        }

        for (int a,b,t,i=1; i<=m; ++i) {
            a = in.nextInt(); b = in.nextInt(); t = in.nextInt();
            if (t==1) {
                e.get(a).add(b);
                et.get(b).add(a);
            } else {
                e.get(a).add(b); e.get(b).add(a);
                et.get(a).add(b); et.get(b).add(a);
            }
        }

        ArrayList<Integer> ord = new ArrayList<Integer>();
        boolean[] vis = new boolean[n+1];
        for (int i=1; i<=n; ++i) 
            if (!vis[i])
                dfs(i, e, vis, ord);

        Arrays.fill(vis, false);
        ArrayList<Integer> ans = new ArrayList<Integer>();
        for (int i=ord.size()-1; i>=0; --i)
            if (!vis[ord.get(i)]) {
                ArrayList<Integer> cc = new ArrayList<Integer>();
                dfs(ord.get(i), et, vis, cc);
                Collections.sort(cc);
                if (cc.size()>ans.size() || 
                        (cc.size()==ans.size() && smaller(cc, ans)))
                    ans = cc;
            }
        System.out.println(ans.size());
        for (int x : ans) 
            System.out.print(x+" ");
    }
}